---
title: "Maria Zakaryan, 2129276"
output: html_document
---

# **Bioinformatics I exam project of RNA-seq analysis.**

## Step 1: Data from Expression Atlas

The experiment chosen from Expression Atlas is:

#### Next Generation Sequencing identifying the dosage compensation state in human endometrial carcinoma and adjacent tissues. [GEO: GSE56087](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56087)

This differential experiment was done using RNA-seq technology by having 9 individuals with 36 samples, where 18 samples were the normal non-cancerous adjacent ones and the other 18 were the endometrial cancerous. Each individual had 2 samples takes from both tissues. The tissues were generated by Illumina 100-nucleotide paired-end sequencing. The main aim of the research was to understand if dosage compensation is completed properly in cancerous cell, or if some genes get over-activated in cancer tissue.

## Step 2: DESeq2

```{r setup, include=FALSE}
knitr::opts_chunk$set( fig.path = "plots/", dev = "png")
```

#### **0)Activating all the necessary libraries for this step**

```{r}
library(DESeq2)
library(ggplot2)
library(pheatmap)
library(RColorBrewer)
library(EnhancedVolcano)
```

#### 1)Setting the working directory

```{r}
setwd("~/Bioinformatics I/Exam project")
```

#### 2)Loading the experiment design data

```{r}
colData <- read.delim("E-GEOD-56087-experiment-design.tsv", header = TRUE, stringsAsFactors = FALSE)
rownames(colData) <- colData[, 1]  #Run IDs as rownames
colData <- colData[, -1]           #Drop the first column
colnames(colData)[3] <- "condition" 
```

```{r}
head(colData)
```

#### 3)Loading the raw counts data

```{r}
counts <- read.delim("E-GEOD-56087-raw-counts.tsv", header = TRUE, stringsAsFactors = FALSE)

counts <- counts[, -2] #remove gene sybmol column

rownames(counts) <- counts[, 1] #setting gene IDs as rownames 
counts <- counts[, -1]
```

```{r}
head(counts)
```

#### 4)Making the DESeq dataset

```{r}
counts <- counts[, rownames(colData)]  #matching columns to colData

dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = colData,
                              design = ~ condition)
```

#### 5)Filtering low expressed genes

```{r}
dds <- dds[rowSums(counts(dds)) > 1, ]
```

#### 6)DESeq pipeline

```{r}
dds <-DESeq(dds)
```

#### 7)PCA plot using vst() function

```{r PCA plot using vst()}
vsd <- vst(dds, blind = FALSE)
DESeq2::plotPCA(vsd, ntop=500, intgroup = 'condition') + theme_bw() + ggtitle("PCA - VST Normalized")
```

#### 8)Heatmap of the sample-to-sample distances

```{r Heatmap sample-to-sample dist}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         main = "Sample-to-sample distances")
```

#### 9)DE results

```{r}
res <- results(dds, contrast = c("condition", "endometrial carcinoma", "normal"))

#shrinks the noisy LFC results towards zero to make more stable and trustworthy results for making the plots more accurate.
res_shrinked <- lfcShrink(dds,
                 contrast = c("condition","endometrial carcinoma","normal"),
                 res = res,
                 type = "ashr")
res
```

#### 10)Volcano plot

```{r Volcano plot}
EnhancedVolcano(res_shrinked,
                lab = rownames(res),
                x = "log2FoldChange",
                y = "pvalue",
                subtitle = "cancer vs normal")
```

#### 11)MA plot

```{r MA plot}
DESeq2::plotMA(res_shrinked, ylim = c(-5, 5)) #MA plot of the results with LFC shrinked for more accuracy
```

#### 12)Heatmap of top 50 DE genes

```{r Heatmap, top 50 DE genes}
vsd <- vst(dds, blind = FALSE)
rld <- rlog(dds, blind = FALSE)

topGenes <- head(order(res$padj), 50)


#using the vsd matrix for the heatmap 
mat_vst <- assay(vsd)[topGenes, ]

pheatmap(mat_vst,
         #show_rownames = FALSE,
         main = "Top 50 DE genes (VST-transformed)")


#using the rlog matrix for the heatmap
mat_rlog <- assay(rld)[topGenes, ]

pheatmap(mat_rlog,
         show_rownames = FALSE,
         main = "Top 50 DE genes (RLOG-transformed)")

```

## Step 3: GO & PEA

#### 0)Necessary libraries for this step

```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
library(AnnotationDbi)
library(ReactomePA)
```

#### 1)Annotating Gene Symbols

```{r}
res$symbol <- mapIds(
    org.Hs.eg.db,
    keys = rownames(res),
    column = "SYMBOL",
    keytype = "ENSEMBL",
    multiVals = "first"
)
```

#### 2)Getting UP regulated and DOWN regulated genes and defining them with "universe="

First and foremost, I set up both up and down regulated genes with the appropriate criteria required, and only after that I combine them to run further analysis.

```{r}
#Up regulated genes
up_genes <- res$symbol[which(res$log2FoldChange > 0 & res$padj < 0.05)]
up_genes <- na.omit(up_genes)

#Down regulated genes
down_genes <- res$symbol[which(res$log2FoldChange < 0 & res$padj < 0.05)]
down_genes <- na.omit(down_genes)

universe_genes <- na.omit(res$symbol)

#Combining genes
comb_genes <- unique(c(up_genes, down_genes))
```

#### 3)Biological process and Molecular factor

Biological process

```{r}
GO_BP <- enrichGO(gene = comb_genes,
                      universe = universe_genes,
                      OrgDb = org.Hs.eg.db,
                      keyType = "SYMBOL",
                      ont = "BP",
                      pAdjustMethod = "BH",
                      qvalueCutoff = 0.05)
```

Molecular function

```{r}
GO_MF <- enrichGO(gene = comb_genes,
                      universe = universe_genes,
                      OrgDb = org.Hs.eg.db,
                      keyType = "SYMBOL",
                      ont = "MF",
                      pAdjustMethod = "BH",
                      qvalueCutoff = 0.05)
```

#### 4)Pathway enrichment analysis

First the Map SYMBOL must be changed to ENTREZ ID for the Enrichment analysis to work in comparison to the previous analyses.

```{r}
PA_entrez <- mapIds(org.Hs.eg.db,
                    keys = comb_genes,
                    column = "ENTREZID",
                    keytype = "SYMBOL",
                    multiVals = "first")
PA_entrez <- na.omit(PA_entrez)
```

Now that the changes are completed the analysis can be ran.

```{r}
PA <-ReactomePA::enrichPathway(gene = PA_entrez,
                                  organism = "human",
                                  pvalueCutoff = 0.05,
                                  readable = TRUE)
```

#### 5)Making dot-plots of the analyses

Biological process

```{r BP dotplot}
dotplot(GO_BP, title = "GO Biological Process")
```

Molecular function

```{r MF dotplot}
dotplot(GO_MF,title = "GO Molecular Function")
```

Enrichment pathway

```{r PA dotplot}
dotplot(PA, title = "Reactome Pathways")
```

# Step 4: GSEA

#### 0)Updating the set of necessary libraries for this step

```{r}
library(msigdbr)
library(dplyr)
library(enrichplot)
```

#### 1)Getting the HALLMARK gene sets

/ Taken from the instruction /

```{r}
h_gene_sets <- msigdbr(species = "Homo sapiens", category = "H")
msigdbr_t2g <- h_gene_sets %>%
  dplyr::distinct(gs_name, ensembl_gene)  # TERM2GENE format

```

#### 2)Creating the ranked list and running GSEA

```{r}
rank <- res$stat
names(rank) <- rownames(res)
rank <- sort(rank, decreasing = TRUE)
```

Running GSEA with Hallmark gene sets

```{r}
gsea <- clusterProfiler::GSEA(geneList = rank,
                              TERM2GENE = msigdbr_t2g,
                              pvalueCutoff = 0.05,
                              verbose = FALSE)

```

#### 3)Plotting Normalized Enrichment Scores

```{r GSEA NES}
gsea_df <- as.data.frame(gsea)

ggplot(gsea_df, aes(NES, reorder(Description, NES),  fill = NES > 0)) +
  geom_bar(stat = "identity") +
  labs(title = "GSEA NES Scores for Hallmark Pathways", x = "", y = "Normalized Enrichment Score") +
  scale_fill_manual(values = c("red", "blue"), labels = c("Down", "Up")) +
  theme_minimal()
```

#### 4)Enrichment plots of the 5 most enriched Hallmark geneset

```{r top 5 enriched genes plots}
top5 <- head(gsea_df$ID, 5)

for (term in top5) {
  print(gseaplot2(gsea, geneSetID = term, title = term))
}
```

#### 5)Creating Heatmap of core genes

```{r}
core_genes <- as.vector(stringr::str_split(string = gsea@result$core_enrichment[1], pattern = "/", simplify = TRUE))
```

```{r Heatmap, core genes}
mat_core <- assay(vsd)[core_genes, ]

pheatmap::pheatmap(mat_core,
                   main = paste("Heatmap of Core Genes:")) 
  
```

# Step 5: FastQC

For accessing the according fastq files the following Bioproject code was obtained from the GEO repository: *PRJNA242387*. With this code the appropriate fastq files used for the FastQC were downloaded and used.\
Specifically fastq files with codes: SRR1200851_1 & SRR1200881_2.

# Step 6: MultiQC

HTML form of the report is present in the folder.

#### Interpretation of the results.

The per base sequence quality showed that vast majority of the read had a score of \> 30 and so located in the green region, which indicates that the quality of score across all bases was good. Per base sequence content, on the other hand, showed some negative results with one of the sequences being at a warning stage and the other not passing the test. The results of that indicate that an alignment process is required here.\
The GC content, which is an other vital criteria for starting any further downstream differential expression analysis, was at a proper range for the first sample, hence passing the test, but had some skewed regions on the other one and raised a warning.\
Moving on to sequence duplication results, it is again observed that the first sample is at a much better condition and passes the test in comparison to the subsequent one that had raised some warnings indicating moderate duplication. That could be a result of bias which is common in low-input RNA-seq.\
Lastly, adapter content, which is also a vital parameter to better understand the sequences. Here both the samples passed the test indicating that no significant adapters were detected, meaning that trimming of those will not be required at this stage.

# Step 7: Data comparison

Comparison of changes in the obtained results from Expression Atlas results: Are there any differences between log2fold change values.

The genes that I took into account for this comparison are CCDC181 and CDHR1.

CCDC181, with geneID ENSG00000117477, has a log2fold change value of -1.32, which matches the results present on Expression Atlas. CDHR1 on the other hand, with geneID ENSG00000148600, has a log2fold change value of 1.4 to which my results did not match and rather showed a value of 2.78.\
From here I concluded that the expression results were far more accurate and reliable for those genes that were down regulated than up regulated, since the former example is a down regulated one whilst the latter is up regulated.

### Answers to more specific questions

#### 1)How many genes are differentially expressed at the thresholds of padj \< 0.05? And padj \< 0.1?

```{r}
#DE genes at padj < 0.05
sum(res$padj < 0.05, na.rm = TRUE)

#DE genes at padj < 0.1
sum(res$padj < 0.1, na.rm = TRUE)
```

#### 2)How many genes at the thresholds of padj \< 0.05 are upregulated ( \> 0) or downregulated?

```{r}
#Total at padj < 0.05
threshold_genes <- res[res$padj < 0.05 & !is.na(res$padj), ]

#upregulated genes (log2FC > 0)
sum(threshold_genes$log2FoldChange > 0)

#downregulated genes (log2FC < 0)
sum(threshold_genes$log2FoldChange < 0)

```

#### 3)Choice one of the GO enrichment results and report how many categories are significant (p.adjust \< 0.05).

```{r}
#Converting GO BP results into a dataframe 
GO_BP_df <- as.data.frame(GO_BP)

#Criteria of interest 
sum(GO_BP_df$p.adjust < 0.05, na.rm = TRUE)
```

#### 4)How many genes are present in the most enriched category of the GO BP enrichment result?

The geneID column contains the **list of genes** associated with that GO term, separated by slashes (/).

```{r}
head(GO_BP_df, 1)
```

After observing them I extract those genes into a variable and count the length of it.

```{r}
top_enrch_list <- strsplit(GO_BP_df$geneID[1], "/")[[1]]
length(top_enrch_list)
```
